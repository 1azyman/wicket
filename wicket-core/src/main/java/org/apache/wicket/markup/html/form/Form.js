/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

function handleDefaultSubmit(event, submitId, keyStroke) {
    const tag = event.target.tagName.toLowerCase();
    const contenteditable = event.target.getAttribute('contenteditable') === 'true';

    // we're interested only in inputs, textareas and contenteditable fields
    if (tag !== 'input' && tag !== 'textarea' && !contenteditable) return;

    // submit for inputs
    if (tag === 'input' && event.which != 13) return;
    // multi-line inputs (textarea, contenteditable)
    if (tag === 'textarea' || contenteditable) {
        switch (keyStroke) {
            case 'ENTER': if (event.which != 13) return;
            case 'CTRL_ENTER': if (!event.ctrlKey || event.which != 13) return; break;
            case 'SHIFT_ENTER': if (!event.shiftKey || event.which != 13) return; break;
            case 'NONE':
            default: return;
        }
    }

    const b = document.getElementById(submitId);
    if (window.getComputedStyle(b).visibility === 'hidden') return;

    event.stopPropagation();
    event.preventDefault();

    if (b != null && b.onclick != null && typeof (b.onclick) != 'undefined') {
        const r = Wicket.bind(b.onclick, b)();
        if (r != false) b.click();
    } else {
        b.click();
    }

    return false;
}
