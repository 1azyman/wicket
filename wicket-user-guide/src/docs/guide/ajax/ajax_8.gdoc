Just like we have seen for links, Lambda expressions bring their benefit also to AJAX components and behaviors. Many of them come with a factory method that accepts a lambda expression to use as callback code. Here is an example for the classic and the lambda way of defining an ajax link:

Classic way:
{code}
AjaxLink<Void> link = new AjaxLink<Void>("linkId") {
	@Override
	public void onClick(AjaxRequestTarget target) {
		label.modelChanging();
		label.setDefaultModelObject("test" + (intx++));
		label.modelChanged();
		target.add(label);
	}
}
{code}

Using lambdas:
{code}
AjaxLink<Void> link = AjaxLink.onClick("linkId", target -> {
	label.modelChanging();
	label.setDefaultModelObject("newString");
	label.modelChanged();
	target.add(label);
});
{code}

Such factory methods can be found also in the utility class @Lambdas@:

{code}
import static org.apache.wicket.lambda.Lambdas.ajaxLink;

...

AjaxLink<Void> link = ajaxLink("linkId", target -> {
	label.modelChanging();
	label.setDefaultModelObject("newString");
	label.modelChanged();
	target.add(label);
});
{code}

The same can be done for behaviors. For example we can create an @AjaxEventBehavior@ in this way:

{code}
  Lambdas.onEvent("click", target -> //some lambda stuf)
{code}


To find out other useful factory methods, you can check @Lambdas@ JavaDoc.
