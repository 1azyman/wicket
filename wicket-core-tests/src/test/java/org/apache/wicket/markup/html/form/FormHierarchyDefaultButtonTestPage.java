package org.apache.wicket.markup.html.form;

import org.apache.wicket.ajax.AjaxRequestTarget;
import org.apache.wicket.ajax.form.AjaxFormSubmitBehavior;
import org.apache.wicket.markup.html.WebPage;
import org.junit.jupiter.api.Assertions;

public class FormHierarchyDefaultButtonTestPage extends WebPage {
    /**
     * For serialization.
     */
    private static final long serialVersionUID = 1L;

    public final Form<Void> parentForm;
    public final Form<Void> childForm;
    public final Button parentSubmit;
    public final Button childSubmit;
    public final TextField<?> parentInput;
    public final TextField<?> childInput;

    /**
     * Construct.
     */
    public FormHierarchyDefaultButtonTestPage() {
        parentForm = new Form<>("parentForm");
        add(parentForm);

        parentInput = new TextField<>("parentInput");
        parentForm.add(parentInput);

        parentSubmit = new Button("parentSubmit");
        parentSubmit.add(new AjaxFormSubmitBehavior(parentForm, "click") {

            @Override
            protected void onSubmit(AjaxRequestTarget target) {
                Assertions.fail("Shouldn't be called");
            }

        });
        parentForm.setDefaultButton(parentSubmit);
        parentForm.add(parentSubmit);

        childForm = new Form<>("childForm");
        parentForm.add(childForm);

        childInput = new TextField<>("childInput");
        childForm.add(childInput);

        childSubmit = new Button("childSubmit");
        childForm.setDefaultButton(childSubmit);
        childForm.add(childSubmit);
    }
}
