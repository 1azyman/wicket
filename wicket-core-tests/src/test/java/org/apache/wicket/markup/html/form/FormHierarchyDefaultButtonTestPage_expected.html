<html>
<head><script type="text/javascript" src="../resource/org.apache.wicket.resource.JQueryResourceReference/jquery/jquery-3.7.1.js"></script>
<script type="text/javascript" src="../resource/org.apache.wicket.ajax.AbstractDefaultAjaxBehavior/res/js/wicket-ajax-jquery.js"></script>
<script type="text/javascript" id="wicket-ajax-debug-enable">
/*<![CDATA[*/
Wicket.Log.enabled=true;
/*]]>*/
</script>
<script type="text/javascript" id="wicket-ajax-base-url">
/*<![CDATA[*/
Wicket.Ajax.baseUrl="wicket/bookmarkable/org.apache.wicket.markup.html.form.FormHierarchyDefaultButtonTestPage?0";
/*]]>*/
</script>
<script type="text/javascript" src="../resource/org.apache.wicket.markup.html.form.Form/Form.js"></script>
<script type="text/javascript">
/*<![CDATA[*/
Wicket.Event.add(window, "domready", function(event) { 
Wicket.Ajax.ajax({"u":"./org.apache.wicket.markup.html.form.FormHierarchyDefaultButtonTestPage?0-1.0-parentForm-parentSubmit","m":"POST","c":"parentSubmit2","f":"parentForm1","sc":"parentSubmit","e":"click"});;
Wicket.Event.add('childForm4', 'keypress', function(event) { handleDefaultSubmit(event, 'childSubmit3','NONE');;});;
Wicket.Event.add('parentForm1', 'keypress', function(event) { handleDefaultSubmit(event, 'parentSubmit2','NONE');;});;
Wicket.Event.publish(Wicket.Event.Topic.AJAX_HANDLERS_BOUND);
;});
/*]]>*/
</script>
</head><body>
<form wicket:id="parentForm" id="parentForm1" method="post" action="./org.apache.wicket.markup.html.form.FormHierarchyDefaultButtonTestPage?0-1.-parentForm"><div id="parentForm1_hf_0" hidden="" class="hidden-fields"></div>

    <input type="text" wicket:id="parentInput" value="" name="parentInput"/>

    <textarea wicket:id="parentTextarea" name="parentTextarea"></textarea>

    <button wicket:id="parentSubmit" name="parentSubmit" id="parentSubmit2"></button>

    <div wicket:id="childForm" id="childForm4">
        <input type="text" wicket:id="childInput" value="" name="childForm:childInput"/>

        <textarea wicket:id="childTextarea" name="childForm:childTextarea"></textarea>

        <button wicket:id="childSubmit" name="childForm:childSubmit" id="childSubmit3"></button>
    </div>
</form>
</body>
</html>
